packages/hooks/src/data/OperationData.ts

```diff
@@ -0,0 +1,114 @@
+import { ApolloClient } from 'apollo-client';
+import { isEqual } from 'apollo-utilities';
+import { invariant } from 'ts-invariant';
+import {
+  ApolloContextValue,
+  parser,
+  DocumentType,
+  operationName
+} from '@apollo/react-common';
+import { DocumentNode } from 'graphql';
+
+import { CommonOptions } from '../types';
+
+export abstract class OperationData<TOptions = any> {
+  private _isMounted: boolean = true;
+  private _options: CommonOptions<TOptions> = {} as CommonOptions<TOptions>;
+  private _previousOptions: CommonOptions<TOptions> = {} as CommonOptions<
+    TOptions
+  >;
+  private _context: ApolloContextValue = {};
+  private _client: ApolloClient<object> | undefined;
+
+  constructor(options?: CommonOptions<TOptions>, context?: ApolloContextValue) {
+    this._options = options || ({} as CommonOptions<TOptions>);
+    this._context = context || {};
+  }
+
+  get isMounted(): boolean {
+    return this._isMounted;
+  }
+
+  set isMounted(mounted: boolean) {
+    this._isMounted = mounted;
+  }
+
+  get options(): CommonOptions<TOptions> {
+    return this._options;
+  }
+
+  set options(newOptions: CommonOptions<TOptions>) {
+    if (!isEqual(this.options, newOptions)) {
+      this.previousOptions = this.options;
+    }
+    this._options = newOptions;
+  }
+
+  get previousOptions(): CommonOptions<TOptions> {
+    return this._previousOptions;
+  }
+
+  set previousOptions(newOptions: CommonOptions<TOptions>) {
+    this._previousOptions = newOptions;
+  }
+
+  get context(): ApolloContextValue {
+    return this._context;
+  }
+
+  set context(newContext: ApolloContextValue) {
+    this._context = newContext;
+  }
+
+  get client(): ApolloClient<object> | undefined {
+    return this._client;
+  }
+
+  set client(newClient: ApolloClient<object> | undefined) {
+    this._client = newClient;
+  }
+
+  public abstract execute(...args: any): any;
+  public abstract afterExecute(...args: any): () => void;
+
+  protected abstract cleanup(): void;
+
+  protected unmount() {
+    this.isMounted = false;
```

Comment:
another unsolicited comment 😅but aren't you missing calling `cleanup` here?

I'm getting a warning when unmounting any component using the hook.

Reply:
@lucasecdb Can you share the exact warning you're seeing?

Reply:
oh, so sorry, I saw the `Query` component in my component stack and though the error was from it. Apparently it's happening because of this custom hook I've created:

```ts
const useLoading = (loading: boolean) => {
  const [mutate] = useMutation<LoadingMutation, LoadingMutationVariables>(
    LOADING_MUTATION
  )

  useEffect(() => {
    mutate({ variables: { loading } })
  }, [loading, mutate])
}
```

and this is the error I'm seeing:

![image](https://user-images.githubusercontent.com/10223856/58392746-d9e56680-8011-11e9-8a37-ea6d3bb4729d.png)

it's still weird that I'm getting this warning, because my component doesn't use any state, it only uses the `graphql` hoc and the hook I've pasted above.

Reply:
So, according to the stack trace of the warning above, the error is coming from the `MutationData.updateResult` function

![image](https://user-images.githubusercontent.com/10223856/58394603-77dd2f00-801a-11e9-84dc-e883bb5e2385.png)